pkgs/hooks_runner/lib/src/build_runner/result.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2025, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// The outcome of an operation that can either succeed with a value of type [T]
+/// or fail with an error of type [E].
+///
+/// This is a sealed class, meaning its implementations are fixed: [Success] and
+/// [Failure].
+///
+/// Modeled after the Rust `Result` type.
+sealed class Result<T, E> {
+  /// Private constructor to prevent direct instantiation.
+  const Result._();
+
+  /// Returns `true` if this is a [Success] instance.
+  bool get isSuccess => this is Success<T>;
+
+  /// Returns `true` if this is a [Failure] instance.
+  bool get isFailure => this is Failure<E>;
+
+  /// Returns this result as a [Success] instance.
+  ///
+  /// Throws if this is a [Failure].
+  Success<T> get asSuccess;
+
+  /// Returns this result as a [Failure] instance.
+  ///
+  /// Throws if this is a [Success].
+  Failure<E> get asFailure;
```

Comment:
Why are the asSuccess/asFailure methods included in the API? From the usage in this PR it's not clear to me that they are needed (see other questions)? `asSuccess` isn't even called at all (if we decide to keep them, it should probably at least be covered by a test).

Reply:
the `is` checks require us to spell out the type arguments unfortunately:

* https://github.com/dart-lang/language/issues/4366

These methods avoid having to spell out these type arguments.